File: Datastructures/Ds Questions/RAIN_WATER_TRAPPING_PROBLEM_SOLUTION.go
/*
Method to calculate the capacity of storing water,
Given array would denote the height of each block
at different steps
*/
package main

import (
	"fmt"
	"math"
)

func main() {

	arr := []int{3, 1, 2, 4, 0, 2, 3, 2}
	fmt.Println(waterTrappingCapacity(arr))
}

func waterTrappingCapacity(heights []int) int {
	size := len(heights)
	left := make([]int, size)
	right := make([]int, size)
	left[0] = heights[0]
	right[size-1] = heights[size-1]
	capacity := 0
	for i := 1; i < size; i++ {
		max := math.Max(float64(left[i-1]), float64(heights[i]))
		left[i] = int(max)
	}
	for i := size - 2; i > 0; i-- {
		max := math.Max(float64(left[i+1]), float64(heights[i]))
		right[i] = int(max)
	}
	for i := 0; i < size; i++ {
		min := math.Min(float64(left[i]), float64(right[i]))
		capacity += int(min) - heights[i]
	}
	return capacity
}

File: Datastructures/RAIN_WATER_TRAPPING_PROBLEM_SOLUTION.java
class Scratch {
    public static void main(String[] args) {
        int[] arr={ 3,1,2,4,0,2,3,2};
        System.out.println(waterTrappingCapacity(arr));
    }
   static int waterTrappingCapacity(int heights[]){
        int size=heights.length;
        int[] left=new int[size];
        int[] right=new int[size];
        left[0]=heights[0];
        right[size-1]=heights[size-1];
        int totalCapacity=0;
        for (int i = 1; i < heights.length; i++) {
            left[i]=Math.max(left[i-1],heights[i]);
        }

        for (int i = size-2; i >=0; i--) {
            right[i]=Math.max(right[i+1],heights[i]);
        }
        for (int i = 0; i < size; i++) {
            totalCapacity+=Math.min(left[i],right[i])-heights[i];
        }
        return totalCapacity;
    }
}